### PR TITLE
Make P11Engine a singleton

### PR DESCRIPTION
**Pull request by OYTIS**

### CODE REVIEW

**Comment by lbonn** [comment]
How did this caused the referenced bug?

File: src/p11engine.h [hunk 1]
```diff
@@ -65,7 +65,10 @@ class P11SlotsWrapper {
 
 class P11Engine {
```
Comment (patriotyk) [inline src/p11engine.h]: derrive it from boost::noncopyable

**Comment by OYTIS** [comment]
@lbonn  Despite its name P11Engine represents a PKCS11-compatible shared library. In destructor it calls PKCS11_CTX_Unload, which deinitializes the library. If the destructor is called while some other P11Engine is alive, the latter will fail because the library was already deinitialized.

**Comment by OYTIS** [comment]
Sorry all, this works, but is non-testable, I'll take another approach.

**Comment by patrickvacek** [comment]
What do you mean, "non-testable"?

**Comment by OYTIS** [comment]
@patrickvacek The problem is P11Engine takes config as a parameter and with a singleton where I only create an instance once in program's lifetime it is doomed to be configured with the first Config I pass to Get() function. It is fine in the real aktualizr, but in tests when a single binary can contain multiple tests with different configs that just doesn't work.

**Comment by cajun-rat** [comment]
Superseded by #541

**Pull request closed without merging**